In [53]:
import pandas as pd

In [54]:
import numpy as np


In [55]:
import os

In [56]:
os.getcwd()

'/content'

In [57]:
df = pd.read_csv("online_retail_II.csv" , encoding= "latin-1 ")

In [59]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

        InvoiceDate  Price  Customer ID         Country  
0  01/12/2009 07:45   6.95      13085.0  United Kingdom  
1  01/12/2009 07:45   6.75      13085.0  United Kingdom  
2  01/12/2009 07:45   6.75      13085.0  United Kingdom  
3  01/12/2009 07:45   2.10      13085.0  United Kingdom  
4  01/12/2009 07:45   1.25      13085.0  United Kingdom

In [60]:
df = df[df["Quantity"]>0]

In [61]:
df = df.dropna(subset= ["Customer ID"])

In [62]:
df


Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

             InvoiceDate  Price  Customer ID         Country  
0       01/12/2009 07:45   6.95      13085.0  United Kingdom  
1       01/12/2009 07:45   6.75      13085.0  United Kingdom  
2       01/12/2009 07:45   6.75      13085.0  United Kingdom  
3       01/12/2009 07:45   2.10      13085.0  United Kingdom  
4       01/12/2009 07:45   1.25      13085.0  United Kingdom  
...                  ...    ...          ...             ...  
525456  09/12/2010 20:01   2.95      17530.0  United Kingdom  
525457  09/12/2010 20:01   3.75      17530.0  United Kingdom  
525458  09/12/2010 20:01   3.75      17530.0  United Kingdom  
525459  09/12/2010 20:01   3.75      17530.0  United Kingdom  
525460  09/12/2010 20:01   1.95      17530.0  United Kingdom  

[407695 rows x 8 columns]

In [63]:
df["Revenue"] = df["Quantity"] * df["Price"]

In [64]:
df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

             InvoiceDate  Price  Customer ID         Country  Revenue  
0       01/12/2009 07:45   6.95      13085.0  United Kingdom    83.40  
1       01/12/2009 07:45   6.75      13085.0  United Kingdom    81.00  
2       01/12/2009 07:45   6.75      13085.0  United Kingdom    81.00  
3       01/12/2009 07:45   2.10      13085.0  United Kingdom   100.80  
4       01/12/2009 07:45   1.25      13085.0  United Kingdom    30.00  
...                  ...    ...          ...             ...      ...  
525456  09/12/2010 20:01   2.95      17530.0  United Kingdom     5.90  
525457  09/12/2010 20:01   3.75      17530.0  United Kingdom     3.75  
525458  09/12/2010 20:01   3.75      17530.0  United Kingdom     3.75  
525459  09/12/2010 20:01   3.75      17530.0  United Kingdom     7.50  
525460  09/12/2010 20:01   1.95      17530.0  United Kingdom     3.90  

[407695 rows x 9 columns]

In [65]:
df_model= df.groupby("Customer ID").agg(
    {
        "Revenue" : "sum",
        "Invoice" : "nunique",
      "Quantity" : "sum"
    }
).rename(columns = {"Invoice" : "Order_nos" , "Revenue" : "total_amount" , "Quantity" : "total_item"})

In [66]:
df_model

total_amount  Order_nos  total_item
Customer ID                                     
12346.0            372.86         11          70
12347.0           1323.32          2         828
12348.0            222.16          1         373
12349.0           2671.14          3         993
12351.0            300.93          1         261
...                   ...        ...         ...
18283.0            641.77          6         336
18284.0            461.68          1         494
18285.0            427.00          1         145
18286.0           1296.43          2         608
18287.0           2345.71          4        1427

[4314 rows x 3 columns]

In [67]:
df_model.shape

(4314, 3)

#Unit of Analysis
One row repsresents one customer number of orders amount of money spent by them and totoal items
#Cleaning
Deleted rows with negative quantity and removed customers with Nan ID
#New Model Set
One row repsresents one customer number of orders amount of money spent by them and totoal items

#Task 2

In [68]:
df_model["repeated_customer"]=(df_model["Order_nos"] > 1).astype(int)

In [69]:
df_model

total_amount  Order_nos  total_item  repeated_customer
Customer ID                                                        
12346.0            372.86         11          70                  1
12347.0           1323.32          2         828                  1
12348.0            222.16          1         373                  0
12349.0           2671.14          3         993                  1
12351.0            300.93          1         261                  0
...                   ...        ...         ...                ...
18283.0            641.77          6         336                  1
18284.0            461.68          1         494                  0
18285.0            427.00          1         145                  0
18286.0           1296.43          2         608                  1
18287.0           2345.71          4        1427                  1

[4314 rows x 4 columns]

#Target Variable
    Repeated Customers
#Def and desc
it represents customer who hav ordered once or more than once all customers with no of orders greater than 1 marked as 1 other with 1 order as labeled as 0

#Type
Binary
#Assmpt
Multiple invoices reflect repeat purchasing behaviour and customer engagement
#Risk
customer ids may be incomplete or inconsistent which could cause repeat customers to be misclassified as one time buyers

#Task 3

In [70]:
customer_spend = df.groupby("Customer ID")["Revenue"].sum()

In [71]:
customer_spend

Customer ID
12346.0     372.86
12347.0    1323.32
12348.0     222.16
12349.0    2671.14
12351.0     300.93
            ...   
18283.0     641.77
18284.0     461.68
18285.0     427.00
18286.0    1296.43
18287.0    2345.71
Name: Revenue, Length: 4314, dtype: float64

In [ ]:
invoice_value = df.groupby(["Customer ID","Invoice"])["Revenue"].sum()
avg_invoice_value = invoice_value.groupby("Customer ID").mean()
avg_invoice_value

Customer ID
12346.0     33.896364
12347.0    661.660000
12348.0    222.160000
12349.0    890.380000
12351.0    300.930000
              ...    
18283.0    106.961667
18284.0    461.680000
18285.0    427.000000
18286.0    648.215000
18287.0    586.427500
Name: Revenue, Length: 4314, dtype: float64

In [72]:
basket_qty= df.groupby(["Customer ID","Invoice"])["Quantity"].sum()
avg_basket_qty = basket_qty.groupby("Customer ID").mean()
avg_basket_qty

Customer ID
12346.0      6.363636
12347.0    414.000000
12348.0    373.000000
12349.0    331.000000
12351.0    261.000000
              ...    
18283.0     56.000000
18284.0    494.000000
18285.0    145.000000
18286.0    304.000000
18287.0    356.750000
Name: Quantity, Length: 4314, dtype: float64

In [73]:
customer_country = df.groupby("Customer ID")["Country"] \
.agg(lambda x: x.mode()[0])


In [74]:
features = pd.DataFrame({
    "customer_spend": customer_spend,
    "avg_order_value": avg_invoice_value,
    "avg_basket_qty": avg_basket_qty,
    "Customer_country": customer_country
})

features.head()


customer_spend  avg_order_value  avg_basket_qty Customer_country
Customer ID                                                                  
12346.0              372.86        33.896364        6.363636   United Kingdom
12347.0             1323.32       661.660000      414.000000          Iceland
12348.0              222.16       222.160000      373.000000          Finland
12349.0             2671.14       890.380000      331.000000            Italy
12351.0              300.93       300.930000      261.000000      Unspecified

In [82]:
num_orders = df.groupby("Customer ID")["Invoice"].nunique()
features["repeat_customer"] = (num_orders > 1).astype(int)

Customer_spend
it represent amount spent by idividual customer
it could help in a way higher spending ones may be repeated orders
one limitation could be that large spend amount could be from a single order rather than repeated orders

avg_order_value
represents the average monetary value per invoice
it may help distinguish bulk buyers from small occasional buyers
limitation can be distorted by one unusually large invoice

avg_basket_qty
represents the average number of items per order
customers with larger baskets may show stronger engagement
limitation can be quantity does not reflect product value differences

ountry
represents the customers primary country based on purchase records
it may help prediction because purchasing behaviour and repeat rates can vary by region
limitation is country differences may reflect delivery or business structure rather than true customer behaviour

In [75]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier


In [83]:
X = features.drop(columns=["repeat_customer"])
y = features["repeat_customer"]

In [84]:
X = pd.get_dummies(X, drop_first=True)
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [85]:
tree = DecisionTreeClassifier(max_depth=4, random_state=42)
tree.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=4, random_state=42)

In [86]:
rf = RandomForestClassifier(
    n_estimators=200,
    max_depth=6,
    random_state=42
)
rf.fit(X_train, y_train)

RandomForestClassifier(max_depth=6, n_estimators=200, random_state=42)

In [87]:
gb = GradientBoostingClassifier(random_state=42)
gb.fit(X_train, y_train)

GradientBoostingClassifier(random_state=42)

In [88]:
print("Decision Tree train score:", tree.score(X_train, y_train))
print("Random Forest train score:", rf.score(X_train, y_train))
print("Boosted Tree train score:", gb.score(X_train, y_train))

Decision Tree train score: 0.9634888438133874
Random Forest train score: 0.9249492900608519
Boosted Tree train score: 1.0


Decision Tree
Training accuracy was higher than validation accuracy indicating mild overfitting this is expected because a single tree can learn specific training patterns

Random Forest
Training and validation scores were closer suggesting improved generalisation due to ensemble averaging across trees

Gradient Boosting
Training accuracy was extremely high with a larger validation gap this suggests higher model complexity and potential overfitting behaviour

In [89]:

X = features.drop(columns=["repeat_customer"])
y = features["repeat_customer"]
X = pd.get_dummies(X, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train, y_train,
    test_size=0.25,
    random_state=42,
    stratify=y_train
)

tree = DecisionTreeClassifier(max_depth=4, random_state=42)
rf = RandomForestClassifier(n_estimators=200, max_depth=6, random_state=42)
gb = GradientBoostingClassifier(random_state=42)

tree.fit(X_train, y_train)
rf.fit(X_train, y_train)
gb.fit(X_train, y_train)
models = {"Tree": tree, "RF": rf, "GB": gb}

for name, m in models.items():
    train_acc = accuracy_score(y_train, m.predict(X_train))
    val_acc = accuracy_score(y_val, m.predict(X_val))
    print(name, "train:", round(train_acc,3),
          "val:", round(val_acc,3),
          "gap:", round(train_acc-val_acc,3))


Tree train: 0.967 val: 0.951 gap: 0.015
RF train: 0.932 val: 0.921 gap: 0.011
GB train: 1.0 val: 0.993 gap: 0.007


Decision Tree
The decision tree achieved training accuracy of 0.967 and validation accuracy of 0.951 with a small gap of 0.015 this indicates mild overfitting which is expected for a single tree model because it can learn specific training patterns the shallow depth constraint likely helped control excessive overfitting

Random Forest
The random forest achieved training accuracy of 0.932 and validation accuracy of 0.921 with a gap of 0.011 the smaller gap compared to the single tree is consistent with ensemble averaging reducing variance and improving generalisation stability

Gradient Boosting
The boosted tree achieved perfect training accuracy 1 and very high validation accuracy 0.993 with a small gap of 0.007 this suggests very strong fit with limited observed overfitting though boosted models are inherently more complex and may still be sensitive to dataset structure

In [91]:
for name, m in models.items():
    test_acc = accuracy_score(y_test, m.predict(X_test))
    print(name, "TEST accuracy:", round(test_acc,3))

Tree TEST accuracy: 0.949
RF TEST accuracy: 0.903
GB TEST accuracy: 0.997


Final evaluation was performed once on the held-out test set to estimate out of sample performance. The test set was not used for model selection or tuning.

Test performance was broadly consistent with validation performance across models, suggesting that model behaviour is reasonably stable. However results should be interpreted cautiously because the target variable is constructed and depends on dataset time coverage and customer ID quality.